grub-core/lib/gpt.c

```diff
@@ -626,10 +626,17 @@ grub_gpt_write_table (grub_disk_t disk, grub_gpt_t gpt,
 		       sizeof (*header));
 
   addr = grub_gpt_sector_to_addr (gpt, grub_le_to_cpu64 (header->header_lba));
+  if (addr == 0)
+    return grub_error (GRUB_ERR_BUG,
+		       "Refusing to write GPT header to address 0x0");
   if (grub_disk_write (disk, addr, 0, sizeof (*header), header))
     return grub_errno;
 
   addr = grub_gpt_sector_to_addr (gpt, grub_le_to_cpu64 (header->partitions));
+  if (addr < 2)
```

Comment:
MBR is always 512 bytes, Primary GPT header is normally 92 bytes but in theory could be larger, up to the disk sector size. So the partition entries table will always then start at the next, sector 2.

If the sector size is something like 4096 then the Primary GPT is at address 8 and primary entries table starts at 16. So this check won't catch the entries table attempting to overwrite the header in that case, just blocking writing to the MBR. But including the sector size in this check adds potentially buggy code to a test that I added just as a last-resort protection against bugs. And two cgpt only supports sector sizes of 512 and because legacy I don't know that hard drives will ever use a logical sector size other than 512 (physical size may be differ but logical is all we care about here).